### DISCUSSION TITLE
Testing Plan for Stacks 2.1 (due December 14 EOD)

### DISCUSSION THREAD

**Discussion by jcnelson** [stacks-network/stacks-core#3242] [General]
# Finish All Testing By December 14 EOD

Tracking sheet: https://docs.google.com/spreadsheets/d/1YoQPwgfJohtLYjqb91vUd6QJV99XRZUV5G1NeV1w1MU/edit#gid=0

Testing Timelines & Estimates: https://www.notion.so/hiropbc/2-1-Testing-Timelines-Estimates-1c9b3ad74d7040f281a17e30a6a51191

This document will be edited as more test items are accepted.

## Blockchain track

- [x] Verify that the `.pox` to `.pox-2` transition works on testnet - @lgalabru - In Progress
   - [x] `.pox-2` instantiates at the expected block height (unit test added for this in #3339) [Regtest integration tests](https://github.com/hirosystems/stacks-2-1-testing/blob/main/tests/integration/pox/stacking/epoch_2_0.spec.ts)
   - [x] All STX stacked in `.pox` are available for spending / re-stacking once `v1_block_height` passes [Regtest integration tests](https://github.com/hirosystems/stacks-2-1-testing/blob/main/tests/integration/pox/stacking/epoch_2_0.spec.ts)
   - [x] Users can re-stack their STX into `.pox-2` [Regtest integration tests](https://github.com/hirosystems/stacks-2-1-testing/blob/main/tests/integration/pox/stacking/epoch_2_0.spec.ts)
   - [x] Once the next reward cycle begins, tokens stacked in `.pox-2` are used to derive the reward cycle.  `.pox` is ignored [Regtest integration tests](https://github.com/hirosystems/stacks-2-1-testing/blob/main/tests/integration/pox/stacking/epoch_2_0.spec.ts)
   - [x] Verify that btc wire format tx for "delegate-stx" does not work before Epoch 2.1 initializes
           Tests end to end: https://github.com/hirosystems/stacks-2-1-testing/tree/main/tests/integration/pox/stacking

- [ ] Verify that new PoX features work on the public 2.1 testnet - @lgalabru | @donpdonp 
  - [X] `stacks-increase` works as expected
  - [x] `stacks-unlock` works as expected
  - [X] stacking to a segwit and taproot address works in pox-2
  - [ ] stacking to a segwit and taproot address receives payout
  - [ ] verify that btc wire format tx for "delegate-stx" works
         ⚠️ Blocked by https://github.com/stacks-network/stacks-blockchain/issues/3465

- [ ] Verify that mining enhancements work on the public 2.1 testnet - @jcnelson 
  - [X] Miners no longer penalized so harshly for missing a block-commit 
  - [X] on-BTC STX-transfer and stacking land on all forks that arise within 6 Bitcoin blocks after they are mined
  - [ ] Miners can mine with a taproot address - @wileyj 

- [X] Verify that revealing a hidden anchor block does not reorg the chain - @jcnelson
  - [X] Verify that a node correctly computes affirmation maps 
  - [X] Verify that the PoX anchor block is the heaviest-confirmed block-commit by BTC weight, even before 2.1
  - [X] Verify that an anchor block missing for more than one reward cycle is marked `a`bsent
  - [X] Verify that the arrival of a missing anchor block _during_ its reward cycle does not reorg the chain
     - [X] Verify that a _running_ node will ignore the anchor block as having been revealed too late
     - [ ] Verify that a _bootstrapping_ node will ignore the anchor block if informed by its bootstrap node
  - [X] Verify that the arrival of an anchor block _after_ its reward cycle does not reorg the chain
  
- [X] Verify that all new Clarity keywords and functions only become available after the transition  - @obycode -  In Progress
   - [x] get pox addrs

- [X] Verify that subnets instantiate on the public 2.1 testnet - @pavitthrap | @gregorycoppola

- [x] Verify that transactions now pay transaction fees before they are executed, such that any transaction with a runtime-caught `CheckError` is mined and the paying account is billed the transaction fee. - @igorsyl / @netrome 
   - e.g. call a contract which tries to invoke a trait implementation within an `at-block` closure, but which is declared _after_ the `at-block` block.
   - test script: https://gist.github.com/igorsyl/7da553fc7cf19b9578abdc97d76eb825

- [x] Trait implementations work correctly - @obycode 
   - [x] Old trait system behavior is preserved
   - [x] New trait system works as intended
   - [x] Cross-epoch contract-calls, with traits defined in different epochs, works as expected (it does in unit tests)
   - [x] All consensus-critical fields in the 2.1 and 2.05 databases are the same, when `next` is sync'ed up to the chain tip prior to launch

- [ ] Build PoC apps to validate the design of new 2.1 features - @mefrem 
   - [ ] Simple oracle built on `from-consensus-buff`.
   - [ ] Simple bridge with `principal-of`, `principal-destruct`, `principal-construct` - In Progress 
   - [ ] Simple mining pool with pay-to-contract, taproot mining, and BTC tx parsing / validation built in `slice` and `get-burn-block-info?`
      - [ ] Can mine a pay-to-contract block
      - [ ] Can mine with taproot
      - [ ] Can parse Bitcoin transactions with `slice`
      - [ ] Can query Bitcoin block header hashes with `get-burn-block-info?`
   - [ ] Simple proof-of-PoX-payout contract built with `get-burn-block-info?` and BTC tx parsing 

- [X] Stacks API works correctly with Epoch 2.1 @donpdonp @zone117x  -  In Progress
   - [x] PoX 2 events get emitted
   - [x] All events we expect from 2.05 get emitted
   - [x] New PoX events get emitted
   - [x] Epoch transition events get emitted

- [ ] Fuzz testing @obycode - In Progress
   - [ ] No inconsistencies between 2.05 and 2.1's Clarity implementation 
   - [ ] Fuzz testing ongoing for 2.1 Clarity 
   
## Client/Wallet track

- [ ] Clients that stack will automatically direct STX to `.pox-2` 
   - [ ] Clients can determine that STX have unlocked after the transition
   - [ ] Clients can determine when the `.pox-2` contract has been instantiated

- [ ] Clients will report the correct locked / unlocked balance after the transition 
   - [ ] At the transition, all STX locked in `.pox` show up as unlocked
   - [ ] STX stacked after the transition show up as locked

- [ ] Clients can permit users to stack to a segwit address in `.pox-2`

## Exchange track - @wileyj | @cuevasm 

- [X] Reach out to all exchanges we know of to inform them that they will need to update their code to stack user STX to `.pox-2`
- [X] Offer them a testnet (or a private testnet) to try out new stacking
- [X] Inform exchanges that they should use an officially-supported client library for `.pox-2` if the are able

---

Please suggest more tests below

For reference, here are the new features being added to 2.1:

* It fixes `principal-of`, which lets you convert a public key in a Bitcoin scriptSig to a Stacks address (it's currently broken on mainnet)
* It adds Clarity built-ins to construct and deconstruct `principal` types from hash160s and version bytes (which is necessary to convert Bitcoin addresses to Stacks principals and back)
* It adds Clarity built-ins to convert between buffers and integers, and it adds a `slice` built-in -- both of which are necessary for efficiently parsing Bitcoin transactions (my `bitcoin.clar` contract certainly works, but it's very costly to use because it must implement all of these functions in Clarity)
* It ensures that any past Bitcoin block hash can be queried, back to system genesis, even if there is no Stacks block chosen in it (right now, you can only get Bitcoin block hashes if there was a Stacks block mined in it.  You can't get a flash block's Bitcoin hash).  This is necessary for validating SPV proofs.
* It exposes PoX address payouts for past blocks, which enables writing Clarity contracts that react to the on-Bitcoin behavior of PoX (such as using PoX to implement a peg-out)
* It adds support for mining and Stacking with taproot addresses, which enables more-complex interactions between Clarity and Bitcoin than you can have today with just segwit and legacy transactions.  This will be important if miners and/or stackers are part of a 2-way peg system.

**Comment by diwakergupta** [comment]
Thanks for putting this together Jude, was on my list for this week -- much appreciated! 🙏🏽

**Comment by saralab** [comment]
Added Owners to the Testing Plan based on discussion today during Architecture meeting and the offsite

**Comment by whoabuddy** [comment]
Could we add a section for Stacking Pool Operators?

After the Gov CAB review of SIP-015 we asked some of the current pool operators to comment on the transition to continuous stacking, and all three were interested in testing on testnet as soon as it's available. [Comments can be seen here.](https://github.com/stacksgov/sips/pull/99#issuecomment-1300833269)

**Reply by jcnelson** [reply]
Sure!  Can some stacking pool operators comment on what kinds of checks they'd like to perform?  We'll have a month between code completion and 2.1 activation, which is plenty of time to verify that new pool software will work.

**Reply by whoabuddy** [reply]
Tagging @yknl @friedger @planbetterHQ here so they can provide more information. If anyone knows handles for other Stacking pool operators feel free to add them as well!

**Reply by planbetterHQ** [reply]
From Planbetter's point of view we are more confortable simply testing the new Pox-2 contract directly in testnet with sufficient time to smooth out possible bugs, ask questions if problems arise and get ready for launch. 

Adding to that, we have a couple of questions:

- Will pooled stackers have to re-stack again (even those that still have multiple cycles left)? The reason I am asking this is because a lot of our users delegate for 12 cycles, and some will still have cycles left when POX2 comes into action. In that case what will happen and what we will have to do as pool operators? I read the SIP and I understand, they **will have to call delegate-stx again**, but if someone could confirm it, I'd be thankful.
- Is there a way to know which pool members have increased their stacking amounts? So we can acknowledge that and issue `delegate-stack-increase`? 
- Is there a way to know which pool members have asked to extend their delegation? Maybe I am confused here since I don't see such method for pooled stackers. Does this mean that this dependes on the `until-burn-ht` function argument on the `delegate-stx` method, which is the same as POX1? If the delegation status of a pool member does not have an untilBurnHeight value set, it means the pool can issue `delegate-stack-extend`. Am I assuming this right?

Thank you

**Reply by jcnelson** [reply]
> Will pooled stackers have to re-stack again (even those that still have multiple cycles left)?

Yes -- all locked tokens will unlock after PoX 2 is live.

> In that case what will happen and what we will have to do as pool operators? I read the SIP and I understand, they will have to call delegate-stx again, but if someone could confirm it, I'd be thankful.

Yes -- the users will each need to call `delegate-stx`, and the pool operator will need to call `delegate-stack-stx` and `stack-aggregation-commit`.

Note that you _additionally_ have the option to call `stack-aggregation-commit-indexed`, which will return an index integer you can pass into `stack-aggregation-increase`.  This new function `stack-aggregation-increase` was added in 2.1 to make it possible to add more STX to an _already-locked-in_ PoX address _without_ needing to first clear the stacking minimum (which `stack-aggregation-commit` forces you to do).  So, if you have some straggling users, you could repeatedly call `stack-aggregation-increase` with their tokens as they delegate them.

> Is there a way to know which pool members have increased their stacking amounts? So we can acknowledge that and issue delegate-stack-increase?

I believe this information is going to be emitted by the node to an event observer, so you'll be able to tell (cc @zone117x).

> Is there a way to know which pool members have asked to extend their delegation? <snip>

You're correct -- as long as the user has still delegated to the pool operator at the current burn height (or if the user didn't specify a burn height), then the pool operator can call `delegate-stack-extend` to keep their STX locked.

**Reply by zone117x** [reply]
> I believe this information is going to be emitted by the node to an event observer, so you'll be able to tell (cc @zone117x).

Yep that's correct. Although we are still waiting on this issue https://github.com/stacks-network/stacks-blockchain/issues/3289 (associated PR https://github.com/stacks-network/stacks-blockchain/pull/3318) to be completed.

**Reply by planbetterHQ** [reply]
Great, thank you for the help @jcnelson, @zone117x, @whoabuddy

**Comment by wileyj** [comment]
@jcnelson i can look into the miner taproot item.

**Comment by saralab** [comment]
Update : 
* Verify that the .pox to .pox-2 transition works on testnet - [@lgalabru](https://github.com/lgalabru) - In Progress 
* Verify that all new Clarity keywords and functions only become available after the transition - In Progress
    * Found a Type Checker regression and [PR](https://github.com/stacks-network/stacks-blockchain/pull/3432) is in review , @obycode is working on property-testing  - @obycode
* Verify that subnets instantiate on the public 2.1 testnet : @gregorycoppola - In Progress
    * Validated  the NFT Use case with the latest 2.1 version 
    * Merging clarity2 changes into subnets code 
* Stacks API works correctly with Epoch 2.1 @Matt - In Progress
    * Working on API updates to match the changes in this [PR](https://github.com/stacks-network/stacks-blockchain/pull/3409) which fixes this [issue](https://github.com/stacks-network/stacks-blockchain/issues/3407)

**Comment by donpdonp** [comment]
[X] `stacks-unlock` test line item should happen at 2022-Dec-14 11am EST

Stacking to taproot address (32 hashbytes) locked for 1 cycle. Unlock burn height 2,411,250
https://explorer.stacks.co/txid/0xc44914e4a03b239b67e7a0d8bf335725b679f4e05becddb2641d1fef3a4e5892?chain=testnet&api=https://2-1-api.testnet.hiro.so

update [2022-Dec-14]: locked STX did unlock after 2,411,250

[] payout to taproot address
btc taproot testnet address: `tb1p3cgmhx7jduegf8u7028he826lat6nectec8mprv8wsfmwjetfccqz0emm3`
to date (2022-Dec-9) no payments have been made to this address. 
Start Burn Height for stack-stx call is 2,409,274, current height is 2,410,510. Maybe its reward slot hasnt been hit yet?

update [2022-Dec-24]: No payouts recorded to [1]. I'm not sure what this means. Perhaps the stacking minimum was above the amount stacked, or some other reason for no payout.
[1] https://blockstream.info/testnet/address/tb1p3cgmhx7jduegf8u7028he826lat6nectec8mprv8wsfmwjetfccqz0emm3

**Comment by netrome** [comment]
I have issues creating a clarity contract that triggers a runtime-caught CheckError. The proposed approach to to invoke a trait implementation within an at-block closure does not work since at-block closures are only allowed to call read-only functions, while traits only specify public functions (that may modify data). i.e. I can't invoke (at-block block-id (contract-call? ...)).

I have tried many other contracts, you can see some of my activity on the test chain [here](https://explorer.stacks.co/address/STJ8F4BTN3YFG60TRTEPCG6QZXAM2A8EMDKCYDF7?chain=testnet&api=https://2-1-api.testnet.hiro.so) - but so far I have yet to figure out a way to trigger a CheckError that's caught at runtime.

@jcnelson I remember you mentioned potential ways to trigger these during our last Blockchain Engineering Weekly meeting, could you write short descriptions here of ideas I can try out?

Any ideas are much appreciated. I'm new to Clarity so I might be missing something obvious. Please respond with any ideas of contracts that could trigger a CheckError at runtime :pray:.

**Reply by netrome** [reply]
Brice saved the day by pointing me to this on Discord https://github.com/stacks-network/stacks-blockchain/blob/next/src/chainstate/stacks/db/transactions.rs#L8921

**Reply by netrome** [reply]
Interesting :thinking: these contracts failed locally with a check-error (on Clarinet console), but the call passes on the testnet chain https://explorer.stacks.co/txid/0x4391167c68a776dbff888f4c3251183b26d2f3832a092b96b05aa3c7e2265178?chain=testnet&api=https://2-1-api.testnet.hiro.so

This is strange. I will look deeper into this tomorrow and see if I can figure out what's going on.

**Reply by netrome** [reply]
I've tried different clarinet versions locally, and they all reproduce the runtime CheckError. However, I'm currently not able to build Clarinet against the Clarity in the `next` branch of the stacks-blockchain. I suspect this is the version running in the 2.1 testnet.

If this is true, I should be able to reproduce the CheckError on `https://stacks-node-api.testnet.stacks.co/` which should not run the `next` Clarity as I understand it. I will try this, and on the side see what I need to compile Clarinet against `next`.

**Reply by netrome** [reply]
Tested with Clarity on `/chore/update-rusqlite (3bc04a8eaa3273d30b239114c4e3e9173c79b1e7)` and I'm still not able to reproduce the passing behavior seen on chain. Just to be sure, I'll try again after merging in the latest `next` branch into `/chore/update-rusqlite`.

**Reply by obycode** [reply]
Make sure to use the right Clarity version when publishing the contracts. Note that this code is valid in Clarity 2 but should trigger the error in Clarity 1.

**Reply by netrome** [reply]
Ah I see. Now I can reproduce the passing behavior locally. I'll publish a contract with version 1 then. Super helpful, thank you!

**Reply by netrome** [reply]
Task failed successfully: I've deployed and executed the contract on-chain with Clarity 1 now, and it fails as expected. ([function invocation](https://explorer.stacks.co/txid/0xc5cc3ce737abe53879efacdc9c073f215846a5cb4989bc722a36e56598d0b81b?chain=testnet&api=https://2-1-api.testnet.hiro.so)). I don't see where in the UI I can verify the actual type of the error, but from my local experiences I know this should be a CheckError. There's a transaction fee in the transaction on the chain.

I will proceed with automating this test case in this PR https://github.com/hirosystems/stacks-2-1-testing/pull/8
